BPM/tempo editor widget

ronso0 opened the pull request:
Closes #12086 and #14611 

Combines the current BPM display/tap stack with a spinbox for tempo.
The Select mode is auto-quit 2 s after the pointer leaves the widget, for Tap mode this happens after 5 s of inactivity.

* **initial state:** invisible overlay with hover effect
* **click:** activate 'select' mode, show Tap and Edit button
* **Tap:** enable tap button
left-click: tap tempo
right-click: tap track BPM
* **Edit:** shows tempo spinbox
enter desired tempo, apply with Enter. Cancel with Esc and when another widget gets keyboard focus.

![bpm-combo-widget](https://github.com/user-attachments/assets/b6b15512-4898-44bc-ad50-1c6238983819)




___

For now this is available  only in LateNight, styled only in PaleMoon.
I'll keep testing it and polish every now and then.

Some more ideas, regarding UX and more functions:
* do we just want to enter numbers or strings like "nnn %" or "3/4" (for polyrhythm sync)?
* should the spinbox/lineedit update when the BPM change?

### TODO
- [ ] style in LateNight Classic
- [ ] Test with touch devices?

ronso0 commented:
Yep, that's the idea.

If track has no BPM, I switches to tempo mode (rate %).

ronso0 commented:
I suggest to ignore this pre-commit complaint..
<img width="492" height="444" alt="image" src="https://github.com/user-attachments/assets/93c0f596-d1ea-487e-9c47-ba3fd3abbcae" />

ronso0 commented:
IMO this is ready for review.
UX is okay, new tolltips in place. Can't tell how this performs with touch devices, ie. how the hide timer behaves when there is no Leave event.

Swiftb0y commented on src/widget/wbpmeditor.cpp:
```diff
@@ -0,0 +1,254 @@
+    pSelectLayout->addWidget(m_pEditSelectButton.get());
+    pSelectLayout->update();
+    pSelectLayout->activate();
+    m_pSelectWidget->setLayout(pSelectLayout);
```
when paired with `make_unique` above
```suggestion
    m_pSelectWidget->setLayout(pSelectLayout.release());
```

Swiftb0y reviewed (commented):
couple thoughts. Yet untested, looking forward to this feature

ronso0 commented on src/skin/legacy/legacyskinparser.cpp:
```diff
@@ -1033,6 +1036,24 @@ QWidget* LegacySkinParser::parseStemLabelWidget(const QDomElement& element) {
+    pBpmEditor->setup(node, *m_pContext);
+    commonWidgetSetup(node, pBpmEditor);
+    // Set tooltips for children
+    const QString tapTooltip = m_tooltips.tooltipForId("tempo_tap_bpm_tap");
+    const QString editTooltip = m_tooltips.tooltipForId("tempo_edit");
```
There's a tooltipId `visual_bpm_edit` for WBpmEditor which is shown if the current widget doesn't supply one (here only the default click listener widget)

<img width="533" height="98" alt="image" src="https://github.com/user-attachments/assets/05d03f50-e01c-4d6b-85ec-403ab0f1adfe" />

ronso0 commented on res/skins/LateNight/style.qss:
```diff
@@ -477,7 +477,7 @@ WSpinny { /*
   qproperty-layoutAlignment: 'AlignHCenter | AlignTop';
 }
 #BpmRateTapContainer {
-  margin: 1px;
+  /*margin: 1px;*/
 }
```
I'm still polishing the skin.
There are still a lot unused/overwritten styles in style.qss

Swiftb0y commented:
Built and tested, works well.

ronso0 commented:
I polished and chnaged the behavior a bit, please test.
Now we can always return to the initial tempo value with Up/Down in the spinbox.
Layout in PaleMoon should also be okay now (both in Full and Compact decks, with and without Sync or rate adjust buttons)

Swiftb0y reviewed (commented):
works very well otherwise. Thank you.

ronso0 commented on src/widget/wbpmeditor.cpp:
```diff
@@ -0,0 +1,325 @@
+    context.hasNodeSelectInt(node, "NumberOfDigits", &m_spinboxDecimals);
+    m_pEditBox->setDecimals(m_spinboxDecimals);
+
+    // TODO Check if we need to set the scale factor for the spinbox
+    // double scaleFactor = context.getScaleFactor();
+    // setScaleFactor(scaleFactor);
+    // m_pEditBox->setsc >setScaleFactor(scaleFactor);
```
Yup, though I didn't test it yet since I'm not sure which scale factor this refers to.

Swiftb0y commented:
>  show above & below value? Ie. wide but small buttons, but could work in LateNight.

Yeah, that was my first idea too. Could work, idk. But maybe this is overpolishing?

ronso0 commented:
I went with +/- buttons side by side at the bottom -- check the gif in the PR description.

Swiftb0y commented:
Looks good. Anything else you wanna do?

ronso0 committed:
add BPM/tempo edit widget WBpmEditor

Combines the current BPM display/tap stack with a spinbox for tempo.

initial state: invisible overlay with over effect
click: show Tap and Edit select buttons

Tap: enable tap button
  tap tempo (left-click)
  tap track BPM (right-click)

Edit: shows tempo spinbox
  enter desired tempo (BPM or rate), apply with Enter

ronso0 committed:
Skins: use new BPM editor widget in LateNight

ronso0 commented on src/skin/legacy/legacyskinparser.cpp:
```diff
@@ -36,6 +36,7 @@
 #include "widget/wbasewidget.h"
 #include "widget/wbattery.h"
 #include "widget/wbeatspinbox.h"
+#include <widget/wbpmeditor.h>
```
Aaaah, that's why pre-commit kept pushing it up to the `<QtIncludes>`
Fixed.

Swiftb0y reviewed (approved):
LGTM. Thank you.
